DotVVM.Diagnostics.StatusPage/StatusViewModel.cs

```diff
@@ -87,25 +91,37 @@ public override Task PreRender()
 
         public List<DotHtmlFileInfo> MasterPages { get; set; }
 
-        public void CompileAll()
+        public async Task CompileAll()
         {
-            var tempMasterPages = new List<DotHtmlFileInfo>();
-            Routes.ForEach(a => BuildView(a, tempMasterPages));
-            Controls.ForEach(a => BuildView(a,tempMasterPages));
+            var tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+
+            var compileTasks = Routes.Select(a => new Task(() => BuildView(a, tempMasterPages))).ToList();
+            compileTasks.AddRange(Controls.Select(a => new Task(() => BuildView(a, tempMasterPages))).ToList());
+
+            compileTasks.ForEach(i => i.Start());
+
+            await Task.WhenAll(compileTasks.ToArray());
+
             while (tempMasterPages.Count > 0)
             {
                 MasterPages.AddRange(tempMasterPages);
-                tempMasterPages.Clear();
-                MasterPages.ForEach(a => BuildView(a, tempMasterPages));
+
+                //.NET Standard 2.1 - better replace with tempMasterPages.Clear()
+                tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+
+                var masterPagesTasks = MasterPages.Select(i => new Task(() => BuildView(i, tempMasterPages))).ToList();
+                masterPagesTasks.ForEach(i => i.Start());
+
+                await Task.WhenAll(masterPagesTasks.ToArray());
             }
         }
 
         public void BuildView(DotHtmlFileInfo file)
         {
-            BuildView(file, MasterPages);
+            BuildView(file, new ConcurrentBag<DotHtmlFileInfo>(MasterPages));
         }
 
-        private void BuildView(DotHtmlFileInfo file, List<DotHtmlFileInfo> tempList)
+        private void BuildView(DotHtmlFileInfo file, ConcurrentBag<DotHtmlFileInfo> tempList)
```

Comment:
I'd make sure to remove duplicate entries before calling this method, especially for the master pages. In case the page compiles successfully, all parallel threads will call `controlFactory.GetControlBuilder` with only one control and get stalled on a lock (that is inside the `GetControlBuilder` to prevent multiple builds of the page. This 1) may reduce the effectiveness of this parallelization, 2) make the .NET think that it could help itself by adding more threads into the pool, which will result in a CPU surge when they will unblock the lock and get to other views.

In case the compilation will fail, the result is not stored and the view is recompiled every time it's requested. So, in case there is a typo in the master page, it will be recompiled for each view on a single thread :)

Reply:
@exi Good point. @RadekCizkovsky Please, can you add the distinct statement?

Reply:
@exyi because the previous comment was meant for you, not me